cartographer/mapping/2d/tsdf_2d.cc

```diff
@@ -0,0 +1,176 @@
+      value_converter(common::make_unique<TSDValueConverter>(
+          truncation_distance, max_weight)),
+      weight_cells_(
+          limits.cell_limits().num_x_cells * limits.cell_limits().num_y_cells,
+          value_converter->getUnknownWeightValue()) {}
+
+TSDF2D::TSDF2D(const proto::Grid2D& proto) : Grid2D(proto) {
+  CHECK(proto.has_tsdf_2d());
+  value_converter = common::make_unique<TSDValueConverter>(
+      proto.tsdf_2d().truncation_distance(), proto.tsdf_2d().max_weight());
+  weight_cells_.reserve(proto.tsdf_2d().weight_cells_size());
+  for (const auto& cell : proto.tsdf_2d().weight_cells()) {
+    CHECK_LE(cell, std::numeric_limits<uint16>::max());
+    weight_cells_.push_back(cell);
+  }
+}
+
+void TSDF2D::SetCell(const Eigen::Array2i& cell_index, const float tsd,
+                     const float weight) {
+  const int flat_index = ToFlatIndex(cell_index);
+  uint16* tsdf_cell = &(*mutable_correspondence_cost_cells())[flat_index];
+  uint16* weight_cell = &weight_cells_[flat_index];
+  if (*tsdf_cell >= value_converter->getUpdateMarker()) {
+    return;
+  }
+  mutable_update_indices()->push_back(flat_index);
+  mutable_known_cells_box()->extend(cell_index.matrix());
+  *tsdf_cell =
+      value_converter->TSDToValue(tsd) + value_converter->getUpdateMarker();
+  *weight_cell = value_converter->WeightToValue(weight);
+  return;
+}
+
+float TSDF2D::GetTSD(const Eigen::Array2i& cell_index) const {
+  if (limits().Contains(cell_index)) {
+    return value_converter->ValueToTSD(
+        correspondence_cost_cells()[ToFlatIndex(cell_index)]);
+  }
+  return value_converter->getMinTSD();
+}
+
+float TSDF2D::GetWeight(const Eigen::Array2i& cell_index) const {
+  if (limits().Contains(cell_index)) {
+    return value_converter->ValueToWeight(
+        weight_cells_[ToFlatIndex(cell_index)]);
+  }
+  return value_converter->getMinWeight();
+}
+
+std::pair<float, float> TSDF2D::GetTSDAndWeight(
+    const Eigen::Array2i& cell_index) const {
+  if (limits().Contains(cell_index)) {
+    int flat_index = ToFlatIndex(cell_index);
+    return std::make_pair(
+        value_converter->ValueToTSD(correspondence_cost_cells()[flat_index]),
+        value_converter->ValueToWeight(weight_cells_[flat_index]));
+  }
+  return std::make_pair(value_converter->getMinTSD(),
+                        value_converter->getMinWeight());
+}
+
+void TSDF2D::GrowLimits(const Eigen::Vector2f& point) {
+  Grid2D::GrowLimits(point,
+                     {mutable_correspondence_cost_cells(), &weight_cells_},
+                     {value_converter->getUnknownTSDValue(),
+                      value_converter->getUnknownWeightValue()});
+}
+
+proto::Grid2D TSDF2D::ToProto() const {
+  proto::Grid2D result;
+  result = Grid2D::ToProto();
+  result.mutable_tsdf_2d()->mutable_weight_cells()->Reserve(
+      weight_cells_.size());
+  for (const auto& cell : weight_cells_) {
+    result.mutable_tsdf_2d()->mutable_weight_cells()->Add(cell);
+  }
```

Comment:
not sure - but maybe you can do something like:
`result.mutable_tsdf_2d()->mutable_weight_cells() = {weight_cells_.begin(), weight_cells_.end()};`

Reply:
Yes, nice.